src/Widgets/FadeBin.vala

```diff
@@ -0,0 +1,150 @@
+public class Tuba.Widgets.FadeBin : Gtk.Widget {
+	const int MAX_HEIGHT = 300;
+	const float FADE_HEIGHT = 125f;
+
+	private Gtk.Widget? _child = null;
+	public Gtk.Widget? child {
+		get { return _child; }
+		set {
+			if (_child != null) _child.unparent ();
+			_child = value;
+			_child.set_parent (this);
+		}
+	}
+
+	~FadeBin () {
+		if (this.child != null) this.child.unparent ();
+	}
+
+	public override Gtk.SizeRequestMode get_request_mode () {
+		if (this.child != null) return this.child.get_request_mode ();
+		return Gtk.SizeRequestMode.HEIGHT_FOR_WIDTH;
+	}
+
+	private bool _reveal = false;
+	public bool reveal {
+		get { return _reveal; }
+		set {
+			if (_reveal != value) {
+				_reveal = value;
+				this.queue_draw ();
+				this.queue_resize ();
+			}
+		}
+	}
+
+	private bool _should_fade = false;
+	private bool should_fade {
+		get { return _should_fade; }
+		set {
+			if (_should_fade != value) {
+				_should_fade = value;
+				this.notify_property ("faded");
+			}
+		}
+	}
+
+	public bool faded {
+		get {
+			return this.should_fade && !reveal;
+		}
+	}
+
+	const Gsk.ColorStop[] GRADIENT = {
+		{ 0f, { 0, 0, 0, 1f } },
+		{ 1f, { 0, 0, 0, 0f } },
+	};
+
+	public override void size_allocate (int width, int height, int baseline) {
+		if (this.child == null) {
+			base.size_allocate (width, height, baseline);
+			this.should_fade = false;
+			return;
+		}
+
+		this.child.allocate (width, height, baseline, null);
+		if (this.reveal) {
+			this.should_fade = false;
+			return;
+		}
+
+		this.should_fade = this.child.get_height () >= MAX_HEIGHT;
```

Comment:
Don't use `get_height ()` like that, it's for the widget itself to call. Use `get_allocated_height ()` (which is deprecated I think?), or better, `height`, since that's what you've just allocated.

EDIT: `child_height` not `height`, if you follow the suggestion above.